src/Common/ARCFileCache.cpp

```diff
@@ -0,0 +1,971 @@
+#include "ARCFileCache.h"
+
+#include <algorithm>
+#include <filesystem>
+#include <IO/Operators.h>
+#include <IO/ReadHelpers.h>
+#include <IO/ReadSettings.h>
+#include <IO/WriteBufferFromFile.h>
+#include <IO/WriteBufferFromString.h>
+#include <pcg-random/pcg_random.hpp>
+#include <Common/SipHash.h>
+#include <Common/hex.h>
+#include <Common/randomSeed.h>
+
+namespace fs = std::filesystem;
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int REMOTE_FS_OBJECT_CACHE_ERROR;
+    extern const int LOGICAL_ERROR;
+}
+
+namespace
+{
+    String keyToStr(const IFileCache::Key & key) { return getHexUIntLowercase(key); }
+}
+
+ARCFileCache::ARCFileCache(const String & cache_base_path_, const FileCacheSettings & cache_settings_)
+    : IFileCache(cache_base_path_, cache_settings_)
+    , max_high_space_size((1.0 - cache_settings_.size_ratio) * cache_settings_.max_size)
```

Comment:
may be makes sense to add strict checking for non zero ratio, as zero ratio might result in very bad behaviour.